### PR TITLE
cache data in mixer for stat var hierarchy 

### PR DESCRIPTION
**Pull request by chejennifer**
- on mixer start up, get and cache the raw svg mapping, the mapping of each sv/svg to its parent svg and the search index

### CODE REVIEW

File: internal/server/server.go [hunk 1]
```diff
@@ -30,11 +30,19 @@ import (
...
 	"cloud.google.com/go/storage"
 	"github.com/datacommonsorg/mixer/internal/base"
+	pb "github.com/datacommonsorg/mixer/internal/proto"
 	"github.com/datacommonsorg/mixer/internal/store"
 	"github.com/datacommonsorg/mixer/internal/translator"
 	"github.com/datacommonsorg/mixer/internal/util"
 )
 
+// Cache holds cached data for the mixer server.
+type Cache struct {
```
Comment (shifucun) [inline internal/server/server.go:40]: comment on what each cache is about?

File: internal/server/server.go [hunk 2]
```diff
@@ -30,11 +30,19 @@ import (
...
+	ParentSvgMap   map[string]string
+	RawSvg         map[string]*pb.StatVarGroupNode
```
Comment (shifucun) [inline internal/server/server.go:42]: Should use this in https://github.com/datacommonsorg/mixer/blob/master/internal/server/statvar_group.go#L118-L134!

File: internal/server/statvar_hierarchy_util.go [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+	return svgResp.StatVarGroups, nil
+}
+
+// GetParentSvgMap gets the mapping of svg/sv id to the parent svg for that svg/sv.
+func GetParentSvgMap(rawSvg map[string]*pb.StatVarGroupNode) map[string]string {
+	parentSvgMap := map[string]string{}
+	for svgID, svgData := range rawSvg {
+		for _, childSvg := range svgData.ChildStatVarGroups {
+			if _, ok := parentSvgMap[childSvg.Id]; !ok {
+				parentSvgMap[childSvg.Id] = svgID
```
Comment (shifucun) [inline internal/server/statvar_hierarchy_util.go:64]: So now only one parent is stored. What is the implication of the parent selection? It came to mind to store all the parents, but looks like for now, only a random random parent would be used?

File: internal/server/statvar_hierarchy_util.go [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+func updateSearchIndex(tokenString string, index map[string]map[string]RankingInfo, nodeID string) {
+	processedTokenString := strings.ToLower(tokenString)
+	processedTokenString = strings.ReplaceAll(processedTokenString, ",", " ")
+	tokenList := strings.Fields(processedTokenString)
+	approxNumPv := len(strings.Split(nodeID, "_"))
+	if approxNumPv == 1 {
+		// when approxNumPv is 1, most likely a non human curated PV and we want them
+		// ranked lower (less approxNumPv, higher ranking)
+		approxNumPv = 30
```
Comment (shifucun) [inline internal/server/statvar_hierarchy_util.go:84]: nit: make this a const

File: internal/server/statvar_hierarchy_util.go [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+	"github.com/datacommonsorg/mixer/internal/util"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/encoding/protojson"
+)
+
+// RankingInfo holds the ranking information for each stat var hierarchy search
+// result.
+type RankingInfo struct {
+	ApproxNumPv int
```
Comment (shifucun) [inline internal/server/statvar_hierarchy_util.go:32]: Add some comment on the struct fields

File: internal/server/statvar_hierarchy_util.go [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+		if _, ok := index[token]; !ok {
+			index[token] = map[string]RankingInfo{}
+		}
+		index[token][nodeID] = rankingInfo
+	}
+}
+
+// GetSearchIndex gets the search index for the stat var hierarchy.
+func GetSearchIndex(rawSvg map[string]*pb.StatVarGroupNode) map[string]map[string]RankingInfo {
+	searchIndex := map[string]map[string]RankingInfo{}
```
Comment (shifucun) [inline internal/server/statvar_hierarchy_util.go:97]: comment on what the key is for each map

File: internal/server/statvar_hierarchy_util_test.go [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+						},
+						{
+							Id:          "sv3",
+							SearchName:  "token4,",
+							DisplayName: "sv4",
+						},
+					},
+				},
+			},
+			map[string]map[string]RankingInfo{
```
Comment (shifucun) [inline internal/server/statvar_hierarchy_util_test.go:117]: I wonder given a search token, how is the RankingInfo used?

File: internal/server/server.go (hunk 1 above)
Comment (shifucun) [reply internal/server/server.go:40]: How about naming with "ParentSvg", "SvgInfo" and "SvgSearchIndex" to make them consistent.

**Review by shifucun — COMMENTED** [review]
looking good! Very clear implementation, just some minor comments

File: internal/server/statvar_hierarchy_util_test.go (hunk 7 above)
Comment (chejennifer) [reply internal/server/statvar_hierarchy_util_test.go:117]: when given a search token, should return the results after having been sorted by rankinginfo .. can mixer return information in a specific order?

File: internal/server/statvar_hierarchy_util.go (hunk 3 above)
Comment (chejennifer) [reply internal/server/statvar_hierarchy_util.go:64]: We were always only choosing a single path so I thought we would only need one parent stored? Would it be better to store all the parents?

File: internal/server/server.go (hunk 2 above)
Comment (chejennifer) [reply internal/server/server.go:42]: I was thinking of doing this in a followup pr

Comment (shifucun) [reply internal/server/server.go:42]: Yep, but you may try locally see if the cache works (it should)

File: internal/server/statvar_hierarchy_util.go (hunk 3 above)
Comment (shifucun) [reply internal/server/statvar_hierarchy_util.go:64]: I think it's better, store a sorted parent and client pick the first one for now.

File: internal/server/statvar_hierarchy_util_test.go (hunk 7 above)
Comment (shifucun) [reply internal/server/statvar_hierarchy_util_test.go:117]: Why not make the struct as map[string][]NewRankingInfo, where NewRankingInfo contains the sv(g) id and RankingInfo?

File: internal/server/server.go (hunk 1 above)
Comment (chejennifer) [reply internal/server/server.go:40]: done

File: internal/server/statvar_hierarchy_util.go (hunk 3 above)
Comment (chejennifer) [reply internal/server/statvar_hierarchy_util.go]: done

File: internal/server/statvar_hierarchy_util.go (hunk 4 above)
Comment (chejennifer) [reply internal/server/statvar_hierarchy_util.go]: done

File: internal/server/statvar_hierarchy_util.go (hunk 5 above)
Comment (chejennifer) [reply internal/server/statvar_hierarchy_util.go:34]: done

File: internal/server/statvar_hierarchy_util.go (hunk 6 above)
Comment (chejennifer) [reply internal/server/statvar_hierarchy_util.go:110]: done

File: internal/server/statvar_hierarchy_util_test.go (hunk 7 above)
Comment (chejennifer) [reply internal/server/statvar_hierarchy_util_test.go:117]: when filtering for tokens, we will need to do an intersection on the results of every token. Map makes it easier to do the intersection

**Review by shifucun — APPROVED** [review]
thanks for update, looking good!

**Pull request merged**